File: dart/a2ui_core/test/conformance/data_model_conformance_test.dart
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import 'package:a2ui_core/a2ui_core.dart';
import 'package:test/test.dart';

import 'conformance_harness.dart';

/// Runs the shared `conformance/core/data_model.yaml` suite against
/// [DataModel].
///
/// Two mappings, both documented in the suite header: `op: delete` maps to
/// `set(path, null)`, since Dart has no `undefined`; `watch` attaches one
/// observer per entry, so a repeated path attaches a second.
void main() {
  final List<Map<String, Object?>> cases = loadConformanceSuite(
    'core/data_model.yaml',
  );

  group('conformance core/data_model.yaml', () {
    test('suite is not empty', () => expect(cases, isNotEmpty));

    for (final testCase in cases) {
      test(testCase['name']! as String, () {
        _runCase(testCase);
      });
    }
  });
}

void _runCase(Map<String, Object?> testCase) {
  // The suite is shared across cases, so deep copy before mutating.
  final model = DataModel(
    _deepCopy(testCase['initial']) ?? <String, Object?>{},
  );
  addTearDown(model.dispose);

  final watched = <_Observer>[];
  final List<Object?> watchPaths =
      (testCase['watch'] as List<Object?>?) ?? const [];
  for (final path in watchPaths) {
    watched.add(_Observer(model, path! as String));
  }

  final steps = testCase['steps']! as List<Object?>;
  for (var i = 0; i < steps.length; i++) {
    final step = steps[i]! as Map<String, Object?>;
    final reason = '${testCase['name']} step $i (${step['op']})';
    for (final observer in watched) {
      observer.resetCount();
    }

    final Object? expectError = step['expect_error'];
    if (expectError != null) {
      expect(
        () => _applyOp(model, step),
        throwsA(_matchesError(expectError as Map<String, Object?>)),
        reason: reason,
      );
      continue;
    }

    _applyOp(model, step, observers: watched, reason: reason);
    _checkNotifications(step, watched, reason);
    _checkWatchedValues(step, watched, reason);
  }
}

void _applyOp(
  DataModel model,
  Map<String, Object?> step, {
  List<_Observer> observers = const [],
  String reason = '',
}) {
  final op = step['op']! as String;
  switch (op) {
    case 'get':
      final Object? actual = model.get(step['path']! as String);
      if (step['expect_absent'] == true) {
        expect(actual, isNull, reason: reason);
      }
      if (step.containsKey('expect_type')) {
        expect(
          actual,
          step['expect_type'] == 'list'
              ? isA<List<Object?>>()
              : isA<Map<Object?, Object?>>(),
          reason: reason,
        );
      }
      if (step.containsKey('expect')) {
        expect(actual, equals(step['expect']), reason: reason);
      }
    case 'set':
      model.set(step['path']! as String, step['value']);
    case 'delete':
      // Dart has no `undefined`; writing null removes the key.
      model.set(step['path']! as String, null);
    case 'dispose':
      model.dispose();
    default:
      fail('Unknown data_model op: $op');
  }
}

void _checkNotifications(
  Map<String, Object?> step,
  List<_Observer> observers,
  String reason,
) {
  final Object? expected = step['expect_notified'];
  if (expected == null) return;
  final List<String> expectedPaths = (expected as List<Object?>).cast<String>();
  final notified = <String>[
    for (final _Observer observer in observers)
      for (var i = 0; i < observer.changeCount; i++) observer.path,
  ];
  expect(
    notified..sort(),
    equals([...expectedPaths]..sort()),
    reason: '$reason: notified observers',
  );
}

void _checkWatchedValues(
  Map<String, Object?> step,
  List<_Observer> observers,
  String reason,
) {
  final Object? expected = step['expect_values'];
  if (expected == null) return;
  (expected as Map<String, Object?>).forEach((path, value) {
    final _Observer observer = observers.firstWhere(
      (o) => o.path == path,
      orElse: () => fail('$reason: $path is not watched'),
    );
    expect(observer.signal.value, equals(value), reason: '$reason: $path');
  });
}

Matcher _matchesError(Map<String, Object?> expectError) {
  final category = expectError['category'] as String?;
  final message = expectError['message'] as String?;
  Matcher matcher = switch (category) {
    'DataError' => isA<A2uiDataError>(),
    'ValidationError' => isA<A2uiValidationError>(),
    'CatalogError' => isA<A2uiCatalogError>(),
    'IntegrityError' => isA<A2uiIntegrityError>(),
    'RecursionError' => isA<A2uiRecursionError>(),
    'ParseError' => isA<A2uiParseError>(),
    'CompileError' => isA<A2uiCompileError>(),
    _ => isA<A2uiError>(),
  };
  if (message != null) {
    matcher = allOf(
      matcher,
      predicate<Object?>((Object? e) {
        return RegExp(message).hasMatch(e.toString());
      }, 'message matching /$message/'),
    );
  }
  return matcher;
}

/// Deep copies plain maps, lists and scalars parsed from a conformance suite.
Object? _deepCopy(Object? value) {
  if (value is Map) {
    return <String, Object?>{
      for (final MapEntry<Object?, Object?> entry in value.entries)
        entry.key.toString(): _deepCopy(entry.value),
    };
  }
  if (value is List) {
    return <Object?>[for (final Object? item in value) _deepCopy(item)];
  }
  return value;
}

/// A single observer attached to one path of a [DataModel].
class _Observer {
  final String path;
  final ReadonlySignal<Object?> signal;
  int _count = 0;

  _Observer(DataModel model, this.path) : signal = model.watch<Object?>(path) {
    signal.subscribe((_) => _count++);
    // preact_signals calls back on subscribe; that is not a change.
    _count = 0;
  }

  int get changeCount => _count;

  void resetCount() => _count = 0;
}

File: renderers/web_core/tests/conformance/data-model.conformance.test.ts
/*
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as assert from 'node:assert';
import {describe, it} from 'node:test';
import {loadConformanceSuite} from './harness.js';
import {DataModel, type DataSubscription} from '../../src/v0_9/state/data-model.js';

/**
 * Runs the shared `conformance/core/data_model.yaml` suite against `DataModel`.
 *
 * One mapping is worth calling out: `op: delete` maps to `set(path, undefined)`,
 * because this implementation removes a key when its value becomes `undefined`.
 *
 * Behaviour that cannot be expressed as a shared, cross-language dataset stays
 * in `src/v0_9/state/data-model.test.ts`.
 */

interface ConformanceStep {
  op: 'get' | 'set' | 'delete' | 'dispose';
  path?: string;
  value?: unknown;
  expect?: unknown;
  expect_absent?: boolean;
  expect_type?: 'list' | 'object';
  expect_values?: Record<string, unknown>;
  expect_notified?: string[];
  expect_error?: {category?: string; message?: string} | string;
}

interface ConformanceCase {
  name: string;
  initial?: Record<string, unknown>;
  watch?: string[];
  steps: ConformanceStep[];
}

interface Observer {
  path: string;
  subscription: DataSubscription<unknown>;
  count: number;
}

function deepCopy<T>(value: T): T {
  return value === undefined ? value : (JSON.parse(JSON.stringify(value)) as T);
}

function errorMessagePattern(expectError: ConformanceStep['expect_error']): RegExp {
  const message = typeof expectError === 'string' ? expectError : (expectError?.message ?? '');
  return new RegExp(message);
}

function applyOp(model: DataModel, step: ConformanceStep, reason: string): void {
  switch (step.op) {
    case 'get': {
      const actual = model.get(step.path!);
      if (step.expect_absent === true) {
        assert.strictEqual(actual, undefined, reason);
      }
      if (step.expect_type !== undefined) {
        assert.strictEqual(Array.isArray(actual) ? 'list' : 'object', step.expect_type, reason);
      }
      if ('expect' in step) {
        assert.deepStrictEqual(actual, step.expect, reason);
      }
      break;
    }
    case 'set':
      model.set(step.path!, step.value);
      break;
    case 'delete':
      // This implementation removes a key when its value becomes `undefined`.
      model.set(step.path!, undefined);
      break;
    case 'dispose':
      model.dispose();
      break;
    default:
      throw new Error(`Unknown data_model op: ${String(step.op)}`);
  }
}

function runCase(testCase: ConformanceCase): void {
  // The suite is parsed once and shared across cases, so the initial data is
  // deep copied before the model mutates it.
  const model = new DataModel(deepCopy(testCase.initial) ?? {});
  const observers: Observer[] = (testCase.watch ?? []).map(path => {
    const observer: Observer = {path, count: 0, subscription: undefined!};
    observer.subscription = model.subscribe(path, () => observer.count++);
    return observer;
  });

  testCase.steps.forEach((step, index) => {
    const reason = `${testCase.name} step ${index} (${step.op})`;
    for (const observer of observers) {
      observer.count = 0;
    }

    if (step.expect_error !== undefined) {
      assert.throws(
        () => applyOp(model, step, reason),
        errorMessagePattern(step.expect_error),
        reason,
      );
      return;
    }

    applyOp(model, step, reason);

    if (step.expect_notified !== undefined) {
      const notified: string[] = [];
      for (const observer of observers) {
        for (let i = 0; i < observer.count; i++) {
          notified.push(observer.path);
        }
      }
      assert.deepStrictEqual(
        notified.sort(),
        [...step.expect_notified].sort(),
        `${reason}: notified observers`,
      );
    }

    for (const [path, expected] of Object.entries(step.expect_values ?? {})) {
      const observer = observers.find(o => o.path === path);
      assert.ok(observer, `${reason}: ${path} is not watched`);
      assert.deepStrictEqual(observer.subscription.value, expected, `${reason}: ${path}`);
    }
  });
}

describe('conformance core/data_model.yaml', () => {
  const cases = loadConformanceSuite<ConformanceCase>('core/data_model.yaml');

  it('suite is not empty', () => {
    assert.ok(cases.length > 0);
  });

  for (const testCase of cases) {
    it(testCase.name, () => runCase(testCase));
  }
});
